json.go

```diff
@@ -103,34 +102,41 @@ func (d Document) WriteOutJSON(w io.Writer) error {
 }
 
 func jsonMarshalDocument(w *bufio.Writer, doc Document, flush bool) error {
-	/*
-		Supported values:
-		bool, for JSON booleans
-		float64, for JSON numbers
-		string, for JSON strings
-		[]interface{}, for JSON arrays
-		map[string]interface{}, for JSON objects
-		nil for JSON null
-	*/
+	// Supported values
+	//
+	// bool                    for JSON booleans
+	// float64                 for JSON numbers
+	// string                  for JSON strings
+	// []interface{}           for JSON arrays
+	// map[string]interface{}  for JSON objects
+	// nil                     for JSON null
+
+	// optionally flush the writer
 	if flush {
 		defer w.Flush()
 	}
 
-	keys := make([]string, 0, len(doc))
-	for k := range doc {
-		keys = append(keys, k)
+	// start serializing the document
+	_, err := w.WriteRune('{')
+	if err != nil {
+		return err
 	}
-	sort.Strings(keys)
-	w.WriteRune('{')
-	for idx, key := range keys {
+
+	for idx, key := range doc.KeysSorted() {
+		// next JSON key/value
 		if idx > 0 {
-			w.WriteRune(',')
+			_, err = w.WriteRune(',')
+			if err != nil {
+				return err
+			}
 		}
-		err := jsonMarshalString(w, key)
+		// write the key
+		err = jsonMarshalString(w, key)
 		if err != nil {
 			return err
 		}
 		w.WriteRune(':')
```

Comment:
```suggestion
		_, err = w.WriteRune(':')
		if err != nil {
			return err
		}
```

We could instead have:
```go
if _, err := w.WriteRune(r); err != nil {
	return err
}
```

but if we do that, we might want to change the other calls to WriteRune in this function as well for consistency.

Another idea to reduce verbosity throughout this file would be to have a custom writer `w` that saves its first error and stops writing if `w.err != nil`, similar to `bufio.Scanner`. That way, we would only need to check errors at the end of the loop.

i.e. 
```
w.WriteRune(...)
...
w.WriteRune(...)
...
if err := w.Err(); err != nil {
	return err
}
```

But I don't think it's necessarily worth the code churn, at least not until we believe that this package will see more code changes in the future (and in any case it shouldn't be part of this PR).

Reply:
This is a WIP

Reply:
Something like this?
```go
// RuneWriter interface ensures the WriteRune method is available
type RuneWriter interface {
	WriteRune(rune) (int, error)
}

// multiErrRuneWriter allows a caller to use this struct multiple times
// and catch or ask for an error at the end either through defer-ing the Err()
// call, or just returning rw.Err()
type multiErrRuneWriter struct {
	writer RuneWriter
	err    error
}

func (w *multiErrRuneWriter) WriteRune(r rune) error {
	// allow calls to "succeed" if we already have an error
	if w.err != nil {
		return nil
	}
	_, err := w.writer.WriteRune(r)
	if err != nil {
		w.err = err
		return nil
	}
	return nil
}

func (w *multiErrRuneWriter) Err() error {
	return w.err
}
```

Reply:
Yes, except that WriteRune doesn't need to return anything. Here's an example of this pattern: https://go.dev/blog/errors-are-values

Reply:
haha - yea, I had that same thought after I posted the initial comment, and quickly edited it with a followup that coincidentally looks almost exactly like the example in the blog post 😄